### ISSUE TITLE
More strict linter

### ISSUE THREAD

**Issue by tzdybal** [celestiaorg/optimint#261]
We need stronger linting. 

@adlerjohn suggested:
 - catching warnings as errors
 - more checks

Ref: #223

**Comment by adlerjohn** [comment]
Should also enable the `staticcheck` linter if it's not enabled currently.

**Comment by liamsi** [comment]
Where is the current config?

**Comment by adlerjohn** [comment]
https://github.com/celestiaorg/optimint/blob/37de59b12db983c0bf2f2d2e6103889c9c85e4a1/.golangci.yml

**Comment by liamsi** [comment]
Ah, OK. https://github.com/celestiaorg/optimint/blob/37de59b12db983c0bf2f2d2e6103889c9c85e4a1/.golangci.yml#L7-L9

More important than warnings as errors is to use the default set of linters here. And maybe disable only some

**Comment by liamsi** [comment]
Could use this as a solid basis IMO: https://github.com/tendermint/tendermint/blob/05d8a0bd05bcdbaefc387244d19a74b31c206240/.golangci.yml#L1-L59

**Comment by liamsi** [comment]
Don't have time to go through each linter individually but a good heuristic is: take the above linked config as a basis. Fix all issues that this catches or explain why some of them should be disabled.

**Comment by tzdybal** [comment]
> Ah, OK.
> 
> https://github.com/celestiaorg/optimint/blob/37de59b12db983c0bf2f2d2e6103889c9c85e4a1/.golangci.yml#L7-L9
> 
> More important than warnings as errors is to use the default set of linters here. And maybe disable only some

Just for clarification: this config adds some linters, not enables only 3 of them.

**Comment by tzdybal** [comment]
```$ golangci-lint linters
Enabled by your configuration linters:
deadcode: Finds unused code [fast: false, auto-fix: false]
errcheck: Errcheck is a program for checking for unchecked errors in go programs. These unchecked errors can be critical bugs in some cases [fast: false, auto-fix: false]
gofmt: Gofmt checks whether code was gofmt-ed. By default this tool runs with -s option to check for code simplification [fast: true, auto-fix: true]
goimports: In addition to fixing imports, goimports also formats your code in the same style as gofmt. [fast: true, auto-fix: true]
gosimple (megacheck): Linter for Go source code that specializes in simplifying a code [fast: false, auto-fix: false]
govet (vet, vetshadow): Vet examines Go source code and reports suspicious constructs, such as Printf calls whose arguments do not align with the format string [fast: false, auto-fix: false]
ineffassign: Detects when assignments to existing variables are not used [fast: true, auto-fix: false]
misspell: Finds commonly misspelled English words in comments [fast: true, auto-fix: true]
staticcheck (megacheck): Staticcheck is a go vet on steroids, applying a ton of static analysis checks [fast: false, auto-fix: false]
structcheck: Finds unused struct fields [fast: false, auto-fix: false]
typecheck: Like the front-end of a Go compiler, parses and type-checks Go code [fast: false, auto-fix: false]
unused (megacheck): Checks Go code for unused constants, variables, functions and types [fast: false, auto-fix: false]
varcheck: Finds unused global variables and constants [fast: false, auto-fix: false]
```

**Comment by adlerjohn** [comment]
> `staticcheck (megacheck): Staticcheck is a go vet on steroids, applying a ton of static analysis checks [fast: false, auto-fix: false]`

Running `staticcheck` locally shows warnings however, fixed only in #262

**Comment by tzdybal** [comment]
> Running `staticcheck` locally shows warnings however,

Yes. This is strange for me.

**Comment by tzdybal** [comment]
`golangci-lint run --exclude-use-default=false` solves this. Apparently SA4011 is excluded by default (among others).

**Comment by liamsi** [comment]
That is why I think it makes sense to explicitly list those that are used. Even if your approach of adding a few is more elegant and minimalistic it would simply save time if someone wants to see which linters are used and which aren't (it would be ideal if the config would simply reflect that and one wouldn't need to check what the defaults are and what not).

**Comment by tzdybal** [comment]
I totally agree with you!  But... this `exclude-use-defaults` is different story - there are default linters, but there are also default exclusions for particular rules. 
To summarize: my plan is to add `exlude-use-defaults=false` to config file and list linters and exclusions explicitly.
